Split generated RST for class reference based on the base type

YuriSizov opened the pull request:
Related to https://github.com/godotengine/godot-proposals/issues/4543.

Splits generated RSTs by using a different file prefix based on the base type of each class (node/resource/other), plus a dedicated group for global scopes (starting with `@`). This works in tandem with the `godot-docs` PR (https://github.com/godotengine/godot-docs/pull/5990) to give this kind of grouping:

![chrome_2022-07-26_18-58-50](https://user-images.githubusercontent.com/11782833/181056401-f94fc2f8-a2cd-482e-8c47-0014159dc46e.png)

https://user-images.githubusercontent.com/11782833/181056420-d76838fa-0673-4496-b558-e72a480f6da0.mp4

YuriSizov committed:
Split generated RST for class reference based on the base type

akien-mga commented:
I like the idea generally, but I see a few issues:

- "Nodes" / "Resources" / "Classes" - the last one is a weird catch-all. Nodes and Resources *are* classes too.
- Doing the split based on adding prefixes means that we break URLs for all existing classes, and we'll have to change `:ref:`s all over the tutorials when referring to a class. Now one would need to remember if the class is a Node, a Resource or another "Class" to use the correct ref prefix.

YuriSizov commented:
> the last one is a weird catch-all

It is, I kept it like that to avoid breaking all the links. Well, I guess the section itself can be called whatever... Objects, maybe? But I think from a casual user's perspective it's quite clear even with the current split.

> we'll have to change `:ref:`s all over the tutorials when referring to a class.

No, internal identifiers for the classes are still the same and all the references should continue to work.

> we break URLs for all existing classes

Yes, that is true. But this is for 4.0+, where a lot of old googled links break already due to renames, right? That said, if we change existing `class_*` prefix in files to something else, then we can generate `class_*` files with redirects or notes for every class.

yedpodtrzitko commented on doc/tools/make_rst.py:
```diff
@@ -941,6 +941,33 @@ def escape_rst(text, until_pos=-1):  # type: (str) -> str
     return text
 
 
+def generate_rst_filename(class_def, state):  # type (ClassDef, State) -> str
```
```suggestion
def generate_rst_filename(class_def, state):  # type: (ClassDef, State) -> str
```

yedpodtrzitko commented on doc/tools/make_rst.py:
```diff
@@ -941,6 +941,33 @@ def escape_rst(text, until_pos=-1):  # type: (str) -> str
+    if class_name.startswith("@"):
+        file_prefix = "global"
+    else:
+        if class_def.inherits:
```
nit: how about merging `else:` from the line above together with this `if` into `elif class_def.inherits:`. It would save one level of indentation which would look nicer.

akien-mga reviewed (approved):
We discussed this in a meeting, the feature makes sense.

We decided to try to keep the existing format for the URL to avoid the issue of having to know how a given class was categorized to guess its URL - several of us tend to go to the class reference for a given class by editing the URL manually in our browser. This might also be relied upon by external tools that want to link to the Godot docs programmatically.

To preserve this feature, we decided to have `make_rst.py` generate the `index.rst` with the subcategories instead of relying on prefixes.

YuriSizov commented:
Changes done, the `index.rst` file is now generated alongside the rest of them, with a hardcoded list of all classes, split into groups as before. I've renamed "Classes" to "Objects", as it makes a bit more sense IMO. The rest should be functionally the same. Looks a little something like this:

![chrome_2022-11-18_00-20-32](https://user-images.githubusercontent.com/11782833/202564617-1d52de4b-2fb1-4b83-8274-74e011a55582.png)

akien-mga reviewed (approved):
It looks amazing! :tada:

akien-mga commented:
Thanks!
